server/src/service.rs

```diff
@@ -217,34 +218,75 @@ impl Service {
             )
             .await?;
 
-        let mut collections = service.collections.lock().await;
-
-        if let Some(secret) = secret {
-            let collection = Collection::new(
-                "login",
-                "default",
-                Arc::clone(&service.manager),
-                Arc::new(Keyring::open("login", secret).await?),
-            );
-            collections.push(collection.path().clone());
-            object_server
-                .at(collection.path().clone(), collection)
+        if secret.is_some() {
+            service
+                .dispatch_default_collection(secret.unwrap(), object_server)
                 .await?;
         }
+        service.dispatch_session_collection(object_server).await?;
 
+        Ok(())
+    }
+
+    pub async fn dispatch_default_collection(
+        &self,
+        secret: Secret,
+        object_server: &zbus::ObjectServer,
+    ) -> Result<(), Error> {
+        let collection = Collection::new(
+            "login",
+            "default",
+            Arc::clone(&self.manager),
+            Arc::new(Keyring::open("login", secret).await?),
+        );
+        self.collections
+            .lock()
+            .await
+            .push(collection.path().clone());
+
+        let keyring = collection.keyring();
+        let n_items = keyring.n_items().await;
+
+        if n_items > 0 {
+            let mut items: Vec<portal::Item> = Vec::with_capacity(n_items);
+            for item in keyring.items().await {
+                items.push(match item {
+                    Ok(item) => item,
+                    Err(err) => return Err(Error::InvalidItem(err)),
+                })
+            }
+
+            for item in items {
+                collection.dispatch_items(item, object_server).await?;
+            }
```

Comment:
why call dispatch for each item separately instead of all of them at once? why not call dispatch directly when you query it from the keyring instead of storing it in a list just for this?

Reply:
Yeah, it was just more readable. Now removed.